File: src/main/kotlin/year2023/day1/p2.py
import regex

numbersStringToInt = {
    "zero": 0,
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9
}

def main():
    f = open('input.txt', 'r')
    sum = 0
    for line in f:
        found = regex.findall(
            "(?:(zero|one|two|three|four|five|six|seven|eight|nine|[1-9]))",
            line,
            overlapped=True)
        first = int(found[0]) if found[0].isdigit() else numbersStringToInt[found[0]]
        second = int(found[-1]) if found[-1].isdigit() else numbersStringToInt[found[-1]]
        sum += first * 10 + second
    print(sum)

main()

File: src/main/kotlin/year2023/day1/p2.kt
package year2023.day1

import java.io.File

val numbersStringToInt: HashMap<String, Int> = hashMapOf(
    "zero" to 0,
    "one" to 1,
    "two" to 2,
    "three" to 3,
    "four" to 4,
    "five" to 5,
    "six" to 6,
    "seven" to 7,
    "eight" to 8,
    "nine" to 9
)

fun trebuchet2(path: String): Int{
    val file = File(path)
    var sum = 0
    file.forEachLine {
        var currentIndex = 0
        val numericValues = mutableListOf<Int>()

        while (currentIndex < it.length) {
            var foundMatch = false
            if(it[currentIndex].isDigit() && numbersToInt.containsKey(it[currentIndex])) {
                numericValues.add(numbersToInt[it[currentIndex]]!!)
                currentIndex++
            }else {
                for (entry in numbersStringToInt) {
                    if (it.startsWith(entry.key, currentIndex)) {
                        numericValues.add(entry.value)
                        currentIndex += entry.key.length-1
                        foundMatch = true
                        break
                    }
                }
                if (!foundMatch) {
                    currentIndex++
                }
            }
        }
        val first = numericValues.first()
        val last = numericValues.last()
        sum += first*10 + last
    }
    return sum
}

fun main(){
    println(trebuchet2("src/main/kotlin/year2023/day1/input.txt"))
}